### PR TITLE
Mirror notifications to Slack

### PR DESCRIPTION
**Pull request by ryanpbrewster**
@camhashemi @leogrim

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/UserValue.scala [hunk 1]
```diff
@@ -156,6 +160,17 @@ object UserValues {
...
+  case class UserValueJsonEncodedHandler[T](override val name: UserValueName, default: T)(implicit format: Format[T]) extends UserValueHandler[T] {
+    def parse(valOpt: Option[String]): T = {
+      val bestEffort = for {
+        str <- valOpt
+        js <- Try(Json.parse(str)).toOption
+        v <- js.asOpt[T]
+      } yield v
+      bestEffort.getOrElse(default)
+    }
+  }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/UserValue.scala:167]: @camhashemi is this garbage?

File: kifi-backend/modules/eliza/app/com/keepit/notify/delivery/NotificationDeliverer.scala [hunk 2]
```diff
@@ -10,23 +11,35 @@ import com.keepit.notify.info.{ NotificationInfoGenerator, StandardNotificationI
...
       notificationJsonFormat.get.extendedJson(infos.head).map { notifJson =>
         recipient match {
-          case UserRecipient(user) => notificationRouter.sendToUser(user, Json.arr("notification", notifJson.json))
+          case UserRecipient(user) =>
+            val extPushes = notificationRouter.sendToUser(user, Json.arr("notification", notifJson.json))
+            val slackPushes = for {
+              vectors <- shoebox.getSlackNotificationVectorsForUser(user)
+              pushes <- FutureHelpers.accumulateRobustly(vectors) {
+                case SlackNotificationVector(slackUserId, slackToken) =>
+                  slack.postToChannel(slackToken, slackUserId.asChannel, SlackMessageRequest.fromKifi(Json.stringify(notifJson.json)))
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/notify/delivery/NotificationDeliverer.scala:40]: This is explicitly garbage. Just for the moment though.

File: kifi-backend/modules/common/app/com/keepit/model/UserValue.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/UserValue.scala:167]: I suppose you only need a `Reads` but other than that looks great. Ideally we will be able to kill the non-typed `JsValueHandler` if this becomes popular :)

File: kifi-backend/modules/eliza/app/com/keepit/notify/delivery/NotificationDeliverer.scala [hunk 3]
```diff
@@ -10,23 +11,35 @@ import com.keepit.notify.info.{ NotificationInfoGenerator, StandardNotificationI
...
+import com.keepit.slack.SlackClient
+import com.keepit.slack.models.{ SlackMessageRequest, SlackNotificationVector }
 import play.api.libs.json.Json
 
 import scala.concurrent.{ ExecutionContext, Future }
 import scala.util.Try
 
 @Singleton
-class WsNotificationDelivery @Inject() (
+class NotificationDeliverer @Inject() (
```
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/notify/delivery/NotificationDeliverer.scala:22]: #EndOfAnEra

File: kifi-backend/modules/eliza/app/com/keepit/notify/delivery/NotificationDeliverer.scala [hunk 4]
```diff
@@ -10,23 +11,35 @@ import com.keepit.notify.info.{ NotificationInfoGenerator, StandardNotificationI
...
 
   def deliver(recipient: Recipient, notif: NotificationWithItems): Future[Unit] = {
     notificationInfoGenerator.generateInfo(recipient, Seq(notif)).flatMap { infos =>
       notificationJsonFormat.get.extendedJson(infos.head).map { notifJson =>
         recipient match {
-          case UserRecipient(user) => notificationRouter.sendToUser(user, Json.arr("notification", notifJson.json))
+          case UserRecipient(user) =>
+            val extPushes = notificationRouter.sendToUser(user, Json.arr("notification", notifJson.json))
+            val slackPushes = for {
+              vectors <- shoebox.getSlackNotificationVectorsForUser(user)
```
Comment (ryanpbrewster) [inline kifi-backend/modules/eliza/app/com/keepit/notify/delivery/NotificationDeliverer.scala:37]: This is cached pretty aggressively.

File: kifi-backend/modules/common/app/com/keepit/model/UserValue.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/model/UserValue.scala:167]: We should fail if there's bad data in there, shouldn't we?

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala [hunk 5]
```diff
@@ -183,6 +183,7 @@ object Shoebox extends Service {
     def internKeep() = ServiceRoute(POST, "/internal/shoebox/database/internKeep")
     def addUsersToKeep(adderId: Id[User], keepId: Id[Keep]) = ServiceRoute(POST, "/internal/shoebox/database/addUsersToKeep", Param("adderId", adderId), Param("keepId", keepId))
     def registerMessageOnKeep() = ServiceRoute(POST, "/internal/shoebox/database/registerMessageOnKeep")
+    def getSlackNotificationVectorsForUser() = ServiceRoute(POST, "/internal/shoebox/database/getSlackNotificationVectorsForUser")
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala:186]: _If_ we do this, let's not make it Slack specific but rather settings-like?

File: kifi-backend/modules/common/app/com/keepit/model/UserValue.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/UserValue.scala:167]: I considered both ways. This has the advantage of being robust against mistakes. Possible danger of losing information, though. Not sure which is best.

**Pull request closed without merging**